### PR TITLE
Algorithms 56 to 60 updated with Histogram Data 

### LINKED ISSUE #17694: HistogramData rollout: algorithms 56-60
HistogramData rollout: algorithms

This issue is part of the HistogramData rollout, see #17641.
Algorithms covered in this issue are:

```
Framework/Algorithms/src/WorkspaceJoiners.cpp
Framework/Crystal/src/AnvredCorrection.cpp
Framework/Crystal/src/CentroidPeaks.cpp
Framework/Crystal/src/FindSXPeaks.cpp
Framework/Crystal/src/IntegratePeakTimeSlices.cpp
```

### PR DESCRIPTION
**Pull request by dtasev**
HistogramData rollout: algorithms

WorkspaceJoiners is now using `setHistogram(...)` instead of copying over each vector, which should signify a significant performance increase, testing script for performance measurement: https://gist.github.com/DimitarTasev/b02979ae42743fcf520eb00c76e6f18a

This issue is part of the HistogramData rollout, see #17641.
Algorithms covered in this issue are:

```
Framework/Algorithms/src/WorkspaceJoiners.cpp
Framework/Crystal/src/AnvredCorrection.cpp
Framework/Crystal/src/CentroidPeaks.cpp
Framework/Crystal/src/FindSXPeaks.cpp
Framework/Crystal/src/IntegratePeakTimeSlices.cpp
```

A performance measurement on `WorkspaceJoiners.cpp` shows improvement in the range of 30-60%, depending on workspace size and iterations, tested on Mantid 3.7 versus current master release build, 
my testing script: https://gist.github.com/DimitarTasev/b02979ae42743fcf520eb00c76e6f18a

Fixes #17694

### CODE REVIEW

File: Framework/Algorithms/src/WorkspaceJoiners.cpp [hunk 1]
```diff
@@ -56,19 +56,14 @@ WorkspaceJoiners::execWS2D(API::MatrixWorkspace_const_sptr ws1,
     auto &outSpec = output->getSpectrum(i);
     const auto &inSpec = ws1->getSpectrum(i);
 
-    // Copy X,Y,E
-    outSpec.setX(XValues);
-    outSpec.dataY() = inSpec.dataY();
-    outSpec.dataE() = inSpec.dataE();
+    outSpec.setHistogram(inSpec.histogram());
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/WorkspaceJoiners.cpp]: I think these changes could give a substantial speedup for all algorithms that are based on `WorkspaceJoiners`. Could you tests that? A quick way to do that if you don't want to blindly implement a performance test is to just run the algorithms for a big enough workspace (using the doctest scripts can be a good starting point, if you adjust the workspace sizes).

File: Framework/Crystal/src/AnvredCorrection.cpp [hunk 2]
```diff
@@ -206,28 +193,46 @@ void AnvredCorrection::exec() {
...
+    Mantid::Kernel::Units::Wavelength wl;
+    auto points = m_inputWS->points(i);
+
+    // Get a reference to the Y's in the output WS for storing the factors
+    auto &Y = correctionFactors->mutableY(i);
+    auto &E = correctionFactors->mutableE(i);
+
+    // share bin boundaries
+    const auto &inSpec = m_inputWS->getSpectrum(i);
+    correctionFactors->setBinEdges(i, inSpec.binEdges());
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/AnvredCorrection.cpp]: Are you sure this is bin edges? Given the `isHist` flag we had previously, I would assume it is unknown, i.e., we should use `sharedX`?

File: Framework/Crystal/src/FindSXPeaks.cpp [hunk 3]
```diff
@@ -139,23 +135,25 @@ void FindSXPeaks::exec() {
...
-      maxY = Y.begin();
-    }
+    auto maxY = (Y.size() > 1)
+                    ? std::max_element(Y.begin() + distmin, Y.begin() + distmax)
+                    : Y.begin();
+    // if (Y.size() > 1) {
+    //   maxY = std::max_element(Y.begin() + distmin, Y.begin() + distmax);
+    // } else {
+    //   maxY = Y.begin();
+    // }
```
Comment (SimonHeybrock) [inline Framework/Crystal/src/FindSXPeaks.cpp]: Remove?

**Review by SimonHeybrock — COMMENTED** [review]
Good, only a few small comments:

File: Framework/Algorithms/src/WorkspaceJoiners.cpp (hunk 1 above)
Comment (DimitarTasev) [inline Framework/Algorithms/src/WorkspaceJoiners.cpp]: I ran a performance measurement, and I see improvement in the range of 30-60%, depending on workspace size and iterations, tested on Mantid 3.7 versus current master release build, however my performance tests have been inaccurate before, so if you'd like to check/confirm the performance difference for yourself I'm leaving my testing script: https://gist.github.com/DimitarTasev/b02979ae42743fcf520eb00c76e6f18a

**Comment by DimitarTasev** [comment]
Restarting build on [Ubuntu](http://builds.mantidproject.org/job/pull_requests-ubuntu/10087/console)

File: Framework/Algorithms/src/WorkspaceJoiners.cpp (hunk 1 above)
Comment (SimonHeybrock) [inline Framework/Algorithms/src/WorkspaceJoiners.cpp]: Nice (have not yet tested myself). Could you add this to the PR description so the information is not lost down here.

**Comment by SimonHeybrock** [comment]
:shipit:

**Pull request merged**